### ISSUE TITLE
Redux persist integration

### ISSUE THREAD

**Issue by arindampradhan** [react-boilerplate/react-boilerplate#2294]
# Integration with redux presist

I was able to do it to some extent, but things like dynamic reducers still confuse me.
I have a working integration model but I am unable to figure out the behavior of the reconciler of redux-persist. It's very weird

Redux persist v5 has 270 issues pending and is very primitive.

## Problems
* Integrate with v5
* Integrate with dynamically injected reducers from react-loadable.
* Integrate with hot-reload

### configureStore.js

`
/**
 * Create the store with dynamic reducers
 */

import { createStore, applyMiddleware, compose } from 'redux';
import { fromJS } from 'immutable';
import { routerMiddleware } from 'react-router-redux';
import thunk from 'redux-thunk';
import createReducer from './reducers';
import immutableTransform from 'redux-persist-transform-immutable';
import { persistStore, persistReducer } from 'redux-persist';
import {
  stateIterator,
  stateGetter,
  stateSetter,
  stateReconciler,
  lastStateInit,
} from 'redux-persist-immutable-state';
import localforage from 'localforage';
import storage from 'redux-persist/lib/storage'; // defaults to localStorage for web and AsyncStorage for react-native
import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2';

const persistConfig = {
  debug: true,
  transforms: [immutableTransform()],
  key: 'root',
  reducerVersion: '1.0',
  stateReconciler: stateReconciler,
  storage: localforage,
  stateIterator: stateIterator,
  stateGetter: stateGetter,
  stateSetter: stateSetter,
  lastStateInit: lastStateInit,
};
const persistedReducer = persistReducer(persistConfig, createReducer());

export default function configureStore(initialState = {}, history) {
  // Create the store with two middlewares
  // 1. X sagaMiddleware: Makes redux-sagas work
  // 2. routerMiddleware: Syncs the location/URL path to the state
  const middlewares = [thunk, routerMiddleware(history)];

  const enhancers = [applyMiddleware(...middlewares)];

  // If Redux DevTools Extension is installed use it, otherwise use Redux compose
  /* eslint-disable no-underscore-dangle, indent */
  const composeEnhancers =
    process.env.NODE_ENV !== 'production' &&
    typeof window === 'object' &&
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
      ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({
          // TODO: Try to remove when `react-router-redux` is out of beta, LOCATION_CHANGE should not be fired more than once after hot reloading
          // Prevent recomputing reducers for `replaceReducer`
          shouldHotReload: false,
        })
      : compose;
  /* eslint-enable */
  if (process.env.NODE_ENV !== 'production') {
    persistConfig.storage = storage;
  }

  const store = createStore(persistedReducer, initialState, composeEnhancers(...enhancers));

  // Extensions
  store.injectedReducers = {}; // Reducer registry

  // Make reducers hot reloadable, see http://mxs.is/googmo
  /* istanbul ignore next */
  if (module.hot) {
    module.hot.accept('./reducers', () => {
      // please use with caution to persist dynamic reducers with redux persist
      // store.replaceReducer(persistReducer(persistConfig, createReducer(store.injectedReducers)));
      store.replaceReducer(store.injectedReducers);
      persistor.persist();
    });
  }
  const persistor = persistStore(store);
  store.persistor = persistor;
  window.store = store;
  return store;
}

`
### reducers.js

`js
import { fromJS } from 'immutable';
import { combineReducers } from 'redux';
`

### I have removed the top level immutable for this.

Let me know if you are able to help it.

I will update another implementation with redux-persist-immutable where I am facing issues with dynimically injected reducers not getting rehyderated.

**Comment by arindampradhan** [comment]
Yes, sure thanks for the help. If you can have any other suggestions to persist redux data with localstorage, because while building any app my data gets remove on refresh. So I don't know if any other out of the box solution to do it. Thanks for building such a wonderful project.

**Comment by julienben** [comment]
Sure! I think my first guess would be to skip the "magic" of redux-persist and just pick a well-maintained browser storage library with no relation to redux (such as [store.js](https://github.com/marcuswestin/store.js) or [localForage](https://github.com/localForage/localForage)). Then wrap your reducers' initialStates in a helper function which checks for persisted state or returns a default initialState. And something similar in the reducer to save the data you want to persist. 

This seems like overkill if you're only doing it to make up for the shortcomings of hot reloading though (vs as an actual feature of your app).

(I didn't build this project, just helping maintain it)

**Comment by satishjhanwer** [comment]
@julienben I was able to integrate `redux-persist-immutable` successfully with RB, not sure if that is what @arindampradhan is asking for.

**Comment by filipecosta01** [comment]
@satishjhanwer could you please share somewhere?

**Comment by satishjhanwer** [comment]
inside `configureStore.js`

add this import statement
`import { persistStore, autoRehydrate } from 'redux-persist-immutable';`

and the below code

```
  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'test') {
    enhancers.push(autoRehydrate());
  }
```

```
  return (done) => {
    // Un-comment following lines to enable persistance to local-storage.
    /* istanbul ignore next */
    if (process.env.NODE_ENV !== 'test') {
      persistStore(store, { blacklist: ['route', 'form'] }, () => done(store));
    }
  };
```

**Comment by Get-Kraken** [comment]
@satishjhanwe can you give more context on where in the file your suggested snippets should go?

**Comment by satishjhanwer** [comment]
@Get-Kraken You need to place the `enhancer.push()` snippet right after the `const enhancer` line and the other snippet at the end of the file replacing this `return store;`. I hope its clear now.

**Comment by Get-Kraken** [comment]
Cool thanks - all working now :)
